src/crimson/osd/osd.cc

```diff
@@ -725,70 +721,30 @@ void OSD::print(std::ostream& out) const
       << superblock.osd_fsid << " [" << superblock.oldest_map
       << "," << superblock.newest_map << "] "
       << "tlb:" << superblock.cluster_osdmap_trim_lower_bound
-      << " pgs:" << get_pg_shard_manager().get_num_pgs()
-      << "}";
-}
-
-void OSD::ShardDispatcher::print(std::ostream& out) const
-{
-  out << "{osd." << osd.superblock.whoami << " "
-      << osd.superblock.osd_fsid << " [" << osd.superblock.oldest_map
-      << "," << osd.superblock.newest_map << "] "
-      << " pgs:" << get_pg_shard_manager().get_num_pgs()
+      << " pgs:" << pg_shard_manager.get_num_pgs()
       << "}";
 }
 
 std::optional<seastar::future<>>
 OSD::ms_dispatch(crimson::net::ConnectionRef conn, MessageRef m)
 {
-  if (get_pg_shard_manager().is_stopping()) {
+  if (pg_shard_manager.is_stopping()) {
     return seastar::now();
   }
-  bool dispatched = true;
-  gate.dispatch_in_background(__func__, *this, [this, conn=std::move(conn),
-                                                m=std::move(m), &dispatched]() mutable {
-    switch (m->get_type()) {
-      case CEPH_MSG_OSD_MAP:
-      case CEPH_MSG_OSD_OP:
-      case MSG_OSD_PG_CREATE2:
-      case MSG_COMMAND:
-      case MSG_OSD_MARK_ME_DOWN:
-      case MSG_OSD_PG_PULL:
-      case MSG_OSD_PG_PUSH:
-      case MSG_OSD_PG_PUSH_REPLY:
-      case MSG_OSD_PG_RECOVERY_DELETE:
-      case MSG_OSD_PG_RECOVERY_DELETE_REPLY:
-      case MSG_OSD_PG_SCAN:
-      case MSG_OSD_PG_BACKFILL:
-      case MSG_OSD_PG_BACKFILL_REMOVE:
-      case MSG_OSD_PG_LEASE:
-      case MSG_OSD_PG_LEASE_ACK:
-      case MSG_OSD_PG_NOTIFY2:
-      case MSG_OSD_PG_INFO2:
-      case MSG_OSD_PG_QUERY2:
-      case MSG_OSD_BACKFILL_RESERVE:
-      case MSG_OSD_RECOVERY_RESERVE:
-      case MSG_OSD_PG_LOG:
-      case MSG_OSD_REPOP:
-      case MSG_OSD_REPOPREPLY:
-      case MSG_OSD_SCRUB2:
-      case MSG_OSD_PG_UPDATE_LOG_MISSING:
-      case MSG_OSD_PG_UPDATE_LOG_MISSING_REPLY:
-      {
-        return shard_dispatchers.local().ms_dispatch(conn, std::move(m));
-      }
-      default:
-      {
-          dispatched = false;
-          return seastar::now();
-      }
-    }
+  auto maybe_ret = do_ms_dispatch(conn, std::move(m));
+  if (maybe_ret == std::nullopt) {
+    return maybe_ret;
+  }
+
+  gate.dispatch_in_background(
+      __func__, *this, [ret=std::move(*maybe_ret)]() mutable {
```

Comment:
has_value() and value() can be used also here.

Reply:
done